### PR TITLE
feat: update signature controller and integrate decoding api

### PR DESCRIPTION
**Pull request by jpuri**
## **Description**

Integrate decoding api.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/3623

## **Manual testing steps**

1. Enable decoding api locally using env variables
2. Submit a permit request
3. Check that decoding api results are present in confirmation object

## **Screenshots/Recordings**
NA

## **Pre-merge author checklist**

- [X] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've completed the PR template to the best of my ability
- [X] I’ve included tests if applicable
- [X] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [X] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by jpuri** [comment]
@SocketSecurity ignore npm/@metamask/signature-controller@21.1.0

**Comment by jpuri** [comment]
@metamaskbot update-policies

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -2000,6 +2000,11 @@ export default class MetamaskController extends EventEmitter {
         ],
       }),
       trace,
+      decodingApiUrl: process.env.DECODING_API_URL,
+      isDecodeSignatureRequestEnabled: () =>
+        this.preferencesController.state.useExternalServices === true &&
```
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:2005]: Do we usually couple API usage to multiple settings, transaction simulation for example doesn't use this?

File: app/scripts/metamask-controller.js [hunk 2]
```diff
@@ -2000,6 +2000,11 @@ export default class MetamaskController extends EventEmitter {
...
+        this.preferencesController.state.useTransactionSimulations &&
```
Comment (matthewwalsh0) [inline app/scripts/metamask-controller.js:2006]: Is this relevant here since this isn't related to transactions?

File: builds.yml [hunk 3]
```diff
@@ -235,6 +235,10 @@ env:
 
   # Used to enable confirmation redesigned pages
   - ENABLE_CONFIRMATION_REDESIGN: ''
+  # Used to enable signature decoding
+  - ENABLE_SIGNATURE_DECODING: ''
+  # Url of the decoding api
```
Comment (matthewwalsh0) [inline builds.yml:240]:
```suggestion
  # URL of the decoding API used to provide additional data from signature requests
```

**Review by matthewwalsh0 — CHANGES_REQUESTED** [review]

File: app/scripts/metamask-controller.js (hunk 1 above)
Comment (jpuri) [reply app/scripts/metamask-controller.js:2005]: here condition is `isDecodeSignatureRequestEnabled` and thus I do not see a downside or adding preferences, env settings here

File: app/scripts/metamask-controller.js (hunk 2 above)
Comment (jpuri) [reply app/scripts/metamask-controller.js:2006]: we use it on signature pages for simulation section and thus plan is to use it here also, we may add another flag for this in future

Comment (matthewwalsh0) [reply app/scripts/metamask-controller.js:2006]: Do we actually rely on the preference in the UI though, I thought this was only used by the transaction controller to query the simulation API?

Comment (jpuri) [reply app/scripts/metamask-controller.js:2006]: Yep we do https://github.com/MetaMask/metamask-extension/blob/68ba6351f710a3a2a18332b075cd6b999baf09dd/ui/pages/confirmations/components/confirm/info/typed-sign/typed-sign.tsx#L58

**Review by matthewwalsh0 — APPROVED** [review]

**Review by pedronfigueiredo — APPROVED** [review]

**Pull request merged**